alf/algorithms/diayn_algorithm.py

```diff
@@ -75,54 +78,52 @@ def __init__(self,
                 last_kernel_initializer=tf.initializers.Zeros())
 
         self._discriminator_net = discriminator_net
-
-    def _encode_skill(self, skill):
-        if tensor_spec.is_discrete(self._skill_spec):
-            return tf.one_hot(indices=skill, depth=self._num_skills)
-        else:
-            return skill
+        self._reward_normalizer = reward_normalizer
 
     def train_step(self, inputs, state, calc_intrinsic_reward=True):
         """
         Args:
-            inputs (tuple):  skill-augmened observation and previous action
-            state (Tensor): state for DIAYN (previous observation)
+            inputs (tuple):  skill-augmened observation and step_type
+            state (Tensor): state for DIAYN (previous skill)
             calc_intrinsic_reward (bool): if False, only return the losses
         Returns:
             TrainStep:
                 outputs: empty tuple ()
-                state: observation
+                state: skill
                 info (DIAYNInfo):
         """
-        observations_aug, _ = inputs
+        observations_aug, step_type = inputs
         observation, skill = observations_aug
+        prev_skill = state
+        batch_size = tf.shape(observation)[0]
 
         if self._encoding_net is not None:
             feature, _ = self._encoding_net(observation)
 
-        prev_feature = state
-        skill_index = tf.cast(skill, tf.int32)
-        skill = self._encode_skill(skill_index)
-
         skill_pred, _ = self._discriminator_net(inputs=feature)
 
-        if tensor_spec.is_discrete(self._skill_spec):
-            skill_discriminate_loss = tf.nn.softmax_cross_entropy_with_logits(
-                labels=skill, logits=skill_pred)
-        else:
-            skill_discriminate_loss = 0.5 * tf.reduce_mean(
-                tf.square(skill - skill_pred), axis=-1)
+        skill_discriminate_loss = tf.nn.softmax_cross_entropy_with_logits(
+            labels=prev_skill, logits=skill_pred)
+
+        valid_masks = tf.cast(
+            tf.not_equal(step_type, StepType.FIRST), tf.float32)
+        valid_masks = tf.reshape(valid_masks, [batch_size])
```

Comment:
why do you need reshape here?

Reply:
valid_masks is of the shape [N, 1] while the loss is of shage [N]